File: poly_apps/flutter_bloom/lib/apps/app_codemart/services_app_codemart/codemart_api_base.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config_app_codemart/api_config_app_codemart.dart';
import '../models_app_codemart/codemart_types.dart';

class CodeMartApiBase {
  final String baseUrl;
  final String namespace;
  String? _token;

  CodeMartApiBase({
    this.baseUrl = ApiConfigAppCodemart.baseUrl,
    this.namespace = ApiConfigAppCodemart.namespace,
  });

  void setToken(String? token) {
    _token = token;
  }

  String? get token => _token;

  Map<String, String> _getHeaders({bool includeContentType = true}) {
    final headers = <String, String>{
      'X-App-Namespace': namespace,
    };

    if (includeContentType) {
      headers['Content-Type'] = 'application/json';
      headers['Accept'] = 'application/json';
    }

    if (_token != null) {
      headers['Authorization'] = 'Bearer $_token';
    }

    return headers;
  }

  Future<ApiResponse<T>> _request<T>({
    required String method,
    required String endpoint,
    Map<String, dynamic>? queryParams,
    Map<String, dynamic>? body,
    T Function(dynamic)? fromJson,
  }) async {
    try {
      String url = '$baseUrl$endpoint';

      if (queryParams != null && queryParams.isNotEmpty) {
        final queryString = Uri(queryParameters: queryParams.map((key, value) => MapEntry(key, value.toString()))).query;
        url = '$url?$queryString';
      }

      final uri = Uri.parse(url);
      http.Response response;

      switch (method.toUpperCase()) {
        case 'GET':
          response = await http.get(uri, headers: _getHeaders());
          break;
        case 'POST':
          response = await http.post(
            uri,
            headers: _getHeaders(),
            body: body != null ? jsonEncode(body) : null,
          );
          break;
        case 'PUT':
          response = await http.put(
            uri,
            headers: _getHeaders(),
            body: body != null ? jsonEncode(body) : null,
          );
          break;
        case 'PATCH':
          response = await http.patch(
            uri,
            headers: _getHeaders(),
            body: body != null ? jsonEncode(body) : null,
          );
          break;
        case 'DELETE':
          response = await http.delete(uri, headers: _getHeaders());
          break;
        default:
          throw Exception('Unsupported HTTP method: $method');
      }

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final jsonResponse = jsonDecode(response.body) as Map<String, dynamic>;

        return ApiResponse<T>(
          success: jsonResponse['success'] as bool? ?? true,
          code: jsonResponse['code'] as int? ?? response.statusCode,
          message: jsonResponse['message'] as String? ?? 'Success',
          data: fromJson != null && jsonResponse['data'] != null ? fromJson(jsonResponse['data']) : null,
          timestamp: jsonResponse['timestamp'] as String? ?? DateTime.now().toIso8601String(),
        );
      } else {
        final jsonResponse = jsonDecode(response.body) as Map<String, dynamic>;
        return ApiResponse<T>(
          success: false,
          code: response.statusCode,
          message: jsonResponse['message'] as String? ?? 'Request failed',
          error: jsonResponse['error'] as String?,
          timestamp: DateTime.now().toIso8601String(),
        );
      }
    } catch (e) {
      return ApiResponse<T>(
        success: false,
        code: 500,
        message: 'API Error: ${e.toString()}',
        error: e.toString(),
        timestamp: DateTime.now().toIso8601String(),
      );
    }
  }

  Future<ApiResponse<T>> get<T>({
    required String endpoint,
    Map<String, dynamic>? queryParams,
    T Function(dynamic)? fromJson,
  }) {
    return _request<T>(
      method: 'GET',
      endpoint: endpoint,
      queryParams: queryParams,
      fromJson: fromJson,
    );
  }

  Future<ApiResponse<T>> post<T>({
    required String endpoint,
    Map<String, dynamic>? body,
    T Function(dynamic)? fromJson,
  }) {
    return _request<T>(
      method: 'POST',
      endpoint: endpoint,
      body: body,
      fromJson: fromJson,
    );
  }

  Future<ApiResponse<T>> put<T>({
    required String endpoint,
    Map<String, dynamic>? body,
    T Function(dynamic)? fromJson,
  }) {
    return _request<T>(
      method: 'PUT',
      endpoint: endpoint,
      body: body,
      fromJson: fromJson,
    );
  }

  Future<ApiResponse<T>> patch<T>({
    required String endpoint,
    Map<String, dynamic>? body,
    T Function(dynamic)? fromJson,
  }) {
    return _request<T>(
      method: 'PATCH',
      endpoint: endpoint,
      body: body,
      fromJson: fromJson,
    );
  }

  Future<ApiResponse<T>> delete<T>({
    required String endpoint,
    T Function(dynamic)? fromJson,
  }) {
    return _request<T>(
      method: 'DELETE',
      endpoint: endpoint,
      fromJson: fromJson,
    );
  }

  Map<String, dynamic> buildQuery({
    Map<String, dynamic> filters = const {},
    int? page,
    int? pageSize,
    String? sort,
    String? order,
  }) {
    final query = <String, dynamic>{};

    filters.forEach((key, value) {
      if (value != null && value != '') {
        query[key] = value;
      }
    });

    if (page != null) query['page'] = page;
    if (pageSize != null) query['pageSize'] = pageSize;
    if (sort != null) query['sort'] = sort;
    if (order != null) query['order'] = order;

    return query;
  }
}

File: poly_apps/nuxt_mainCopy/apps/app_codemart/services_app_codemart/codemart-api-base.ts
import type { FetchOptions } from 'ofetch';
import { $fetch } from 'ofetch';

export interface ApiResponse<T = any> {
  code: number;
  message: string;
  data: T;
  timestamp?: string;
}

export interface ApiError {
  code: string;
  message: string;
  details?: Record<string, any>;
}

export interface PaginationParams {
  page?: number;
  pageSize?: number;
  sort?: string;
  order?: 'asc' | 'desc';
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export abstract class CodeMartApiBase {
  protected baseUrl: string;
  protected namespace: string = 'codemart';

  constructor(baseUrl: string = '/api/codemart') {
    this.baseUrl = baseUrl;
  }

  protected getHeaders(includeContentType: boolean = true): Record<string, string> {
    const headers: Record<string, string> = {
      'X-App-Namespace': this.namespace,
    };

    if (includeContentType) {
      headers['Content-Type'] = 'application/json';
    }

    return headers;
  }

  protected async request<T = any>(
    endpoint: string,
    options: FetchOptions = {}
  ): Promise<ApiResponse<T>> {
    const url = `${this.baseUrl}${endpoint}`;

    try {
      const response = await $fetch<ApiResponse<T>>(url, {
        ...options,
        headers: {
          ...this.getHeaders(),
          ...(options.headers || {}),
        },
      });

      if (!response || response.code !== 0) {
        throw new Error(response?.message || 'Unknown error');
      }

      return response;
    } catch (error) {
      if (error instanceof Error) {
        throw new Error(`API Error: ${error.message}`);
      }
      throw error;
    }
  }

  protected async get<T = any>(
    endpoint: string,
    query?: Record<string, any>
  ): Promise<ApiResponse<T>> {
    const params = new URLSearchParams();
    if (query) {
      Object.entries(query).forEach(([key, value]) => {
        if (value !== null && value !== undefined) {
          params.append(key, String(value));
        }
      });
    }

    const queryString = params.toString();
    const url = queryString ? `${endpoint}?${queryString}` : endpoint;

    return this.request<T>(url, {
      method: 'GET',
    });
  }

  protected async post<T = any>(
    endpoint: string,
    data?: Record<string, any>
  ): Promise<ApiResponse<T>> {
    return this.request<T>(endpoint, {
      method: 'POST',
      body: data,
    });
  }

  protected async put<T = any>(
    endpoint: string,
    data?: Record<string, any>
  ): Promise<ApiResponse<T>> {
    return this.request<T>(endpoint, {
      method: 'PUT',
      body: data,
    });
  }

  protected async delete<T = any>(endpoint: string): Promise<ApiResponse<T>> {
    return this.request<T>(endpoint, {
      method: 'DELETE',
    });
  }

  protected async patch<T = any>(
    endpoint: string,
    data?: Record<string, any>
  ): Promise<ApiResponse<T>> {
    return this.request<T>(endpoint, {
      method: 'PATCH',
      body: data,
    });
  }

  protected async postFormData<T = any>(
    endpoint: string,
    formData: FormData
  ): Promise<ApiResponse<T>> {
    return this.request<T>(endpoint, {
      method: 'POST',
      body: formData,
      headers: this.getHeaders(false),
    });
  }

  protected buildQuery(
    filters: Record<string, any> = {},
    pagination: PaginationParams = {}
  ): Record<string, any> {
    const query: Record<string, any> = {};

    Object.entries(filters).forEach(([key, value]) => {
      if (value !== null && value !== undefined && value !== '') {
        query[key] = value;
      }
    });

    if (pagination.page) query.page = pagination.page;
    if (pagination.pageSize) query.pageSize = pagination.pageSize;
    if (pagination.sort) query.sort = pagination.sort;
    if (pagination.order) query.order = pagination.order;

    return query;
  }
}

export default CodeMartApiBase;
